src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResultBuilder.cs

```diff
@@ -95,14 +101,51 @@ public TimeSpan GetAnalyzerExecutionTime(DiagnosticAnalyzer analyzer)
             }
         }
 
-        internal ImmutableArray<DiagnosticAnalyzer> GetPendingAnalyzers(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        private HashSet<DiagnosticAnalyzer>? GetCompletedAnalyzersForFile_NoLock(SourceOrAdditionalFile filterFile, bool syntax)
+        {
+            if (filterFile.SourceTree is { } tree)
+            {
+                var completedAnalyzersByTree = syntax ? _completedSyntaxAnalyzersByTree : _completedSemanticAnalyzersByTree;
+                if (completedAnalyzersByTree.TryGetValue(tree, out var completedAnalyzersForTree))
+                {
+                    return completedAnalyzersForTree;
+                }
+            }
+            else if (_completedSyntaxAnalyzersByAdditionalFile.TryGetValue(filterFile.AdditionalFile!, out var completedAnalyzersForFile))
```

Comment:
i'd prefer `else if (filterFile.AdditionalFile != null)`  then an `else` that throws for not having source-tree/additional file. up to you.